(DOCSP-21897): Context (#1866)
mongodb/docs-realm: source/functions/context.txt

     - A built-in HTTP client.

   * - :ref:`context.request <context-request>`
     - Describes the incoming request that triggered a function call.

   * - :ref:`context.services <context-services>`
     - Exposes client objects that can access :ref:`data sources
       <data-sources>` and :ref:`services <services>`.

   * - :ref:`context.user <context-user>`
     - Describes the authenticated :ref:`user
       <users-and-authentication>` that initiated the request.

   * - :ref:`context.values <context-values>`
     - Contains static global :ref:`values <values-and-secrets>`.

.. _context-functions:

Call a Function (``context.functions``)
---------------------------------------

You can call any :ref:`function <functions>` in your application
through the ``context.functions`` interface.

.. method:: context.functions.execute()

   Calls a specific function and returns the result.

   .. code-block:: javascript
      

>       context.functions.execute(functionName, args...)

Reviewer: ```suggestion
      context.functions.execute(functionName, ...args)
```
nit: i'd do like this so it's like a `...` JS spread operator

Author: +1 good catch

   
   .. list-table::
      :header-rows: 1
      :widths: 20 10 70

      * - Parameter
        - Type
        - Description

      * - ``functionName``
        - string
        - The name of the function.

      * - ``args``...
        - mixed
        - A variadic list of arguments to pass to the function. Each
          function parameter maps to a separate, comma-separated
          argument.

   .. example::
      
      .. code-block:: javascript
         
         // difference: subtracts b from a using the sum function
         exports = function(a, b) {
             return context.functions.execute("sum", a, -1 * b);

>          };

Reviewer: had to think abt this one for a sec 🤓

   
   Possible values:

   .. include:: /includes/list-of-environment-names.rst
   
   .. example::
      
      .. code-block:: javascript
         
         exports = async function() {
           switch(context.environment.tag) {
             case "": {
               return "There is no current environment"
             }
             case "development": {
               return "The current environment is development"
             }
             case "testing": {
               return "The current environment is testing"
             }
             case "qa": {
               return "The current environment is qa"
             }
             case "production": {
               return "The current environment is production"
             }
           }
         };

.. method:: context.environment.values
   
   An object where each field maps the name of an :ref:`environment value
   <environment-values>` to its value in the current environment.
   
   .. example::
      
      .. code-block:: javascript
         
         exports = async function() {
           const baseUrl = context.environment.values.baseUrl
         };

.. _context-services:

> Connect to a Data Source or Service (``context.services``)
> ----------------------------------------------------------

Reviewer: ```suggestion
Connect to a MongoDB Data Source or Service (``context.services``)
----------------------------------------------------------------------
```

worth explicitly calling out

> You can access a client for a linked cluster, data lake, or third-party

Reviewer: ```suggestion
You can access a client for a linked MongoDB Atlas cluster, Atlas Data Lake, or third-party
```

> You can access a client for a linked cluster, data lake, or third-party
> service through the ``context.services`` interface.

Reviewer: ```suggestion
You can access a client for a linked MongoDB Atlas cluster or Data Lake through the ``context.services`` interface. You can also access third-party services, although this feature is deprecated. 
```

Reviewer: smthn like this. good to explicitly call out connecting to atlas.

.. method:: context.services.get()

   Gets a service client for the specified service or ``undefined`` if
   no such service exists.
   
   .. code-block:: javascript
      
      context.services.get(serviceName)

   .. list-table::
      :header-rows: 1
      :widths: 20 10 70

      * - Parameter
        - Type
        - Description

      * - ``serviceName``
        - string
        - The name of the linked cluster, data lake, or service.
          
          .. tip:: MongoDB Service Names
             
             Linked :ref:`data sources <mongodb-service>` created by your app use
             one of the following default names:

             - Cluster: ``mongodb-atlas``

>              - Data Lake: ``mongodb-datalake``

Reviewer: not nec for this PR, but for the coming work on using realm w serverless instances, should we add a default name for serverless instances here?

Author: Yup I'll make sure to include it!

           const http = context.services.get("myHttpService");
           // Call the HTTP service's get() action
           try {
             const response = await http.get({ url: "https://www.mongodb.com" });
             return response.body.text()
           } catch(err) {
             // You might get an error if:
             // - you passed invalid arguments
             // - the service's rules prevent the action
             console.error(err)
           }
         };

.. _context-request:

Get Request Metadata (``context.request``)
------------------------------------------

You can access information about the incoming request with the
:ref:`context.request <context-request>` interface.

.. method:: context.request
   
   An object that contains information about the HTTP request that
   caused the function to execute.
   
   .. code-block:: javascript
   
      {
         "remoteIPAddress": <string>,
         "requestHeaders": <object>,
         "webhookUrl": <string>,
         "httpMethod": <string>,
         "rawQueryString": <string>,
         "httpReferrer": <string>,
         "httpUserAgent": <string>,
         "service": <string>,
         "action": <string>

>       }

Reviewer: is the request body not part of this?

Author: Nope - if you're in e.g. an HTTPS endpoint handler then the request body is in the `request` argument that's passed in to the endpont function. Intuitive!

Some subset of this full doc is available for any given function execution depending on the context.

Reviewer: ok, do you think it's worth adding a lil note for talking about accessing body from HTTPS endpoint handler?
